common/src/main/java/tachyon/network/protocol/RPCMessageEncoder.java

```diff
@@ -42,9 +42,9 @@ protected void encode(ChannelHandlerContext ctx, RPCMessage in, List<Object> out
     long bodyBytes = 0;
     DataBuffer payload = null;
 
-    if (in.hasPayload()) {
-      payload = in.getPayloadDataBuffer();
-      bodyBytes = payload.getLength();
+    if (in instanceof PayloadMessage) {
+      payload = ((PayloadMessage)in).getPayloadDataBuffer();
+      bodyBytes = (payload == null) ? 0 : payload.getLength();
```

Comment:
I am actually kind of ambivalent about this change.  I can imagine PayloadMessage having more functionality like setPayload, compressPayload etc. Or we might have additional interface/subclasses for Messages like compressedMessages.  @gpang, what do you think? is this what you had in mind for the jira?